src/CustomFunctionsTrait.php

```diff
@@ -0,0 +1,270 @@
+
+    /**
+     * Registers a custom function.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $code     The code to register the function with.
+     * @param callable          $function The function to register.
+     *
+     * @throws InvalidArgumentException If function is not a callable.
+     */
+    protected function _registerCustomFunction($code, $function)
+    {
+        $code = $this->_normalizeFunctionCode($code);
+
+        if (!is_callable($function, true)) {
+            throw $this->_createInvalidArgumentException($this->__('Function being registered with code "%1$s" must be a callable', $code), null, null, $function);
+        }
+
+        $this->customFunctions[$code] = $function;
+    }
+
+    /**
+     * Registers multiple custom functions in bulk.
+     *
+     * @since [*next-version*]
+     *
+     * @param callable[]|Traversable $map A map of codes to function names.
+     *
+     * @throws InvalidArgumentException If the map, or a code or function in it
+     *                                  is illegal.
+     */
+    protected function _registerCustomFunctions($map)
+    {
+        if (!is_array($map) && !($map instanceof Traversable)) {
+            throw $this->_createInvalidArgumentException($this->__('Invalid custom functions map'), null, null, $map);
+        }
+
+        foreach ($map as $_code => $_function) {
+            $this->_registerCustomFunction($_code, $_function);
+        }
+    }
+
+    /**
+     * Unregisters a custom function.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $code The code of the function to unregister.
+     *
+     * @throws OutOfRangeException      If no function with the specified code is registered.
+     * @throws InvalidArgumentException If function code is illegal.
+     */
+    protected function _unregisterCustomFunction($code)
+    {
+        $code = $this->_normalizeFunctionCode($code);
+
+        if (!isset($this->customFunctions[$code])) {
+            throw $this->_createOutOfRangeException($this->__('No function with this code is defined'), null, null, $code);
+        }
+
+        unset($this->customFunctions[$code]);
+    }
+
+    /**
+     * Calls a custom function by its registered code.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $code The code of the function to call.
+     * @param array             $args The arguments to pass to the function.
+     *
+     * @throws DomainException          If the function is not callable
+     * @throws InvalidArgumentException If the list of arguments or the function
+     *                                  code is illegal.
+     * @throws OutOfRangeException      If no function with the specified code is defined.
+     * @throws DomainException          If the function is not callable.
+     *
+     * @return mixed The result of the function call.
+     */
+    protected function _callCustomFunction($code, $args = [])
```

Comment:
Similarly, why not just type hint `$args` as `array`?

Reply:
By analogy, there could be an implementation with a different mechanism for args passing, such as a `Traversable & ContainerInterface`. Thoughts?

Reply:
Not sure what my thoughts are if I'm being honest. Lately I've been thinking about what is best: adapting from the outside or from the inside?

Say there's some component that needs to work with an array. Should it accept multiple types and adapt them into an array? Or should the consumer, who knows what the API is and knows that the component needs an array, be responsible for adapting the data that it has into an array?

I'm not sure there's a "correct" answer. It sounds situational.

In this case, because you are using `call_user_func_array`, you **have** to use an array. Leaving it open for implementation is meaningless, since this is a trait and the method is not part of a public API. If this were an interface method or an abstract method, I would leave this open.

Reply:
An implementation can adapt something that is not an array, and turn it into an array. That's why I mentioned `Traversable & ContainerInterface`. But I would say that I much prefer adapting from the outside. And by that I mean accepting something more concrete (simplifies the implementation), and when something wants to look like something else, it implements something else. If it really *is* that similar, then the adaptation is just a forward in 99% of cases.

Reply:
Agree. What about our current case? Does it really matter that it's left unhinted? What are your thoughts about my argument about this being a trait and not part of a public API?

Reply:
The thoughts about this being a trait is that it's still a contract, even if not an API.

The other side is that, like I wrote before, adaptation from the outside is better IMHO. It's still possible to produce a different implementation of the interface. By tightening the type, we could increase maintainability and readability. So, I think I'm gona do it.